cmd/kubelet/app/server.go

```diff
@@ -412,13 +431,249 @@ func run(s *options.KubeletServer, kcfg *KubeletConfig) (err error) {
+	}
+	if err = crypto.WriteKeyToPath(keyPath, keyData); err != nil {
+		if err := os.Remove(certPath); err != nil {
+			glog.Warningf("Cannot clean up the certificate file %q: %v", certPath, err)
+		}
+		return "", "", fmt.Errorf("unable to write key file %q: %v", keyPath, err)
+	}
+
+	return certPath, keyPath, nil
+}
+
+// getNodeName returns the node name according to the cloud provider
+// if cloud provider is specified. Otherwise, returns the host name of the node.
+func getNodeName(s *options.KubeletServer) (string, error) {
+	var err error
+	var cloud cloudprovider.Interface
+
+	if s.CloudProvider != kubeExternal.AutoDetectCloudProvider {
+		cloud, err = cloudprovider.InitCloudProvider(s.CloudProvider, s.CloudConfigFile)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	hostName := nodeutil.GetHostname(s.HostnameOverride)
+	if cloud != nil {
+		instances, ok := cloud.Instances()
+		if !ok {
+			return "", fmt.Errorf("failed to get instances from cloud provider")
+		}
+		return instances.CurrentNodeName(hostName)
+	}
+	return hostName, nil
+}
+
+// requestClientCertificate will create a certificate signing request and send it to API server,
+// then it will watch the object's status, once approved by API server, it will return the API
+// server's issued certificate (pem-encoded). If there is any errors, or the watch timeouts,
+// it will return an error.
+// If the existingKeyData is empty, a new private key will be generated to create the certificate
+// signing request.
+func requestClientCertificate(client unversionedcertificates.CertificateSigningRequestInterface, existingKeyData []byte, nodeName string) (certData []byte, keyData []byte, err error) {
+	subject := &pkix.Name{
+		Organization: []string{"system:nodes"},
+		CommonName:   fmt.Sprintf("system:node:%s", nodeName),
+	}
+
+	csr, keyData, err := utilcertificates.NewCertificateRequest(existingKeyData, subject, nil, nil)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to generate certificate request: %v", err)
+	}
+
+	req, err := client.Create(&certificates.CertificateSigningRequest{
+		TypeMeta:   unversioned.TypeMeta{Kind: "CertificateSigningRequest"},
+		ObjectMeta: api.ObjectMeta{GenerateName: "csr-"},
+
+		// Username, UID, Groups will be injected by API server.
+		Spec: certificates.CertificateSigningRequestSpec{Request: csr},
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot create certificate signing request: %v", err)
+
+	}
+
+	// Make a default timeout = 3600s
+	var defaultTimeoutSeconds int64 = 3600
+	resultCh, err := client.Watch(api.ListOptions{
+		Watch:          true,
+		TimeoutSeconds: &defaultTimeoutSeconds,
+		// Label and field selector are not used now.
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot watch on the certificate signing request: %v", err)
+	}
+
+	var status certificates.CertificateSigningRequestStatus
+	ch := resultCh.ResultChan()
+
+	for {
+		event, ok := <-ch
+		if !ok {
+			break
```

Comment:
a watch can drop for a couple reasons: 1) timeout, or 2) too many other server-side events have happened since the resourceVersion the watch was opened on. Now that I think about it, though, I'm not sure the watch window will matter here if you're watching a cluster-scoped resource, and are not specifying a starting resource version